apiserver/facades/agent/keyupdater/authorisedkeys.go

```diff
@@ -52,100 +36,176 @@ var _ KeyUpdater = (*KeyUpdaterAPI)(nil)
+
+		if tag.Kind() != names.MachineTagKind {
+			results[i].Error = apiservererrors.ParamsError(
+				params.CodeTagKindNotSupported,
+				"tag %q unsupported, can only accept tags of kind %q",
+				tag, names.MachineTagKind,
+			)
+			continue
+		}
+
 		if !canRead(tag) {
-			results[i].Error = apiservererrors.ServerError(apiservererrors.ErrPerm)
+			results[i].Error = apiservererrors.ParamsError(
+				params.CodeUnauthorized,
+				"does not have permission to read authorised keys for %q",
+				tag,
+			)
 			continue
 		}
-		// 2. Check entity exists
-		if _, err := api.state.FindEntity(tag); err != nil {
-			if errors.Is(err, errors.NotFound) {
-				results[i].Error = apiservererrors.ServerError(apiservererrors.ErrPerm)
-			} else {
-				results[i].Error = apiservererrors.ServerError(err)
-			}
+
+		machineId := coremachine.Name(tag.Id())
+		keysWatcher, err := api.keyUpdaterService.WatchAuthorisedKeysForMachine(ctx, machineId)
+		switch {
+		case errors.Is(err, errors.NotValid):
+			results[i].Error = apiservererrors.ParamsError(
+				params.CodeMachineInvalidID,
+				"invalid machine name %q",
+				machineId,
+			)
 			continue
+		case err != nil:
+			// We don't understand this error. At this stage we consider it an
+			// internal server error and bail out of the call completely.
+			return params.NotifyWatchResults{}, fmt.Errorf(
+				"cannot watch authorised keys for machine %q: %w",
+				machineId, err,
+			)
 		}
-		// 3. Watch for changes
-		watch := api.model.WatchForModelConfigChanges()
-		// Consume the initial event.
-		if _, ok := <-watch.Changes(); ok {
-			results[i].NotifyWatcherId = api.resources.Register(watch)
-		} else {
-			err = watcher.EnsureErr(watch)
+
+		notifyWatcher, err := watcher.Normalise[[]string](keysWatcher)
+		if err != nil {
+			return params.NotifyWatchResults{}, fmt.Errorf(
+				"converting machine %q authorised keys watcher to notify watcher: %w",
+				machineId, err,
+			)
+		}
+
+		results[i].NotifyWatcherId, _, err = internal.EnsureRegisterWatcher[struct{}](
+			ctx, api.watcherRegistery, notifyWatcher,
+		)
+		if err != nil {
+			return params.NotifyWatchResults{}, fmt.Errorf(
+				"registering machine %q authorised keys watcher: %w",
+				machineId, err,
+			)
 		}
-		results[i].Error = apiservererrors.ServerError(err)
 	}
 	return params.NotifyWatchResults{Results: results}, nil
 }
 
 // AuthorisedKeys reports the authorised ssh keys for the specified machines.
-// The current implementation relies on global authorised keys being stored in
-// the model config and also the Juju system ssh key coming from the controllers
-// config. This will change as new user management and authorisation
-// functionality is added.
-func (api *KeyUpdaterAPI) AuthorisedKeys(ctx context.Context, arg params.Entities) (params.StringsResults, error) {
+// The current implementation fetches all keys on the model that have been
+// granted for use on a machine.
+// The following param error codes can be expected:
```

Comment:
The doc string for the watch method should have this too right?

Reply:
yes it should......

Reply:
Done & Done